### PR TITLE
Allow bootstrapping on 10.14.5 w/ perl 5.18.4

### PR DESCRIPTION
**Pull request by nieder**

### CODE REVIEW

**Comment by nieder** [comment]
bb608f2 and 4be81e2 are needed for bootstrapping.

e13ad45 is necessary, but I'm not sure sufficient , to create a new dist for 10.14.5.

**Comment by akhansen** [comment]
Seems sensible to me.

**Comment by akhansen** [comment]
Bootstrap worked for me

**Fionna:fink hansen$ /tmp/sw/bin/fink list perl**
Scanning package description files..........
Information about 319 packages read in 1 seconds.
 p   perl5184-core                 [virtual package]
 p   perlio-eol-pm51               [virtual package]
 p   perlobjcbridge-               [virtual package]
 i   system-perl      5.18.4-1     [virtual package representing perl]
 p   system-perl5184               [virtual package]

File: perlmod/Fink/Bootstrap.pm [hunk 1]
```diff
@@ -245,9 +245,12 @@ GCC_MSG
...
 			"guarantees.");
 		$distribution = "10.13";
-	} elsif ($host =~ /^i386-apple-darwin18\.[0-6]/) {
+	} elsif ($host =~ /^i386-apple-darwin18\.[0-5]/) {
 		&print_breaking("This system is supported and tested.");
 		$distribution = "10.14";
+	} elsif ($host =~ /^i386-apple-darwin18\.[6]/) {
+		&print_breaking("This system is supported and tested.");
+		$distribution = "10.14.5";
 	} elsif ($host =~ /^i386-apple-darwin18\./) {
```
Comment (PovlAbrahamsen) [inline perlmod/Fink/Bootstrap.pm]: Shouldn't line 259 now read `$distribution = "10.14.5";`, to ensure that the newer distribution is used for future updates?

**Comment by mrvacbob** [comment]
I tried bootstrapping on 10.14.6 with this and get a failure:
```
/sw/bin/fink -y install gettext-bin libgettext8-dev libiconv-dev libncurses5
Use 'fink reinstall fink' to switch distributions
from 10.14.5 to 10.14.
'install' operation not permitted.
### execution of /tmp/fink.4nAiX failed, exit code 29
phase compiling: dpkg-1.10.21-1248 failed
```

**Comment by nieder** [comment]
@mrvacbob was this a fresh bootstrap, or an update of an existing install?

**Comment by mrvacbob** [comment]
This is a fresh bootstrap from the `nieder-5.18.4-bootstrap` branch on 18G29g.

**Comment by nieder** [comment]
What does 10.14.6 report with `uname -a` ?

**Comment by mrvacbob** [comment]
`Darwin asimac.local 18.6.0 Darwin Kernel Version 18.6.0: Tue May  7 22:54:55 PDT 2019; root:xnu-4903.270.19.100.1~2/RELEASE_X86_64 x86_64`

File: perlmod/Fink/Bootstrap.pm (hunk 1 above)
Comment (akhansen) [reply perlmod/Fink/Bootstrap.pm]: Yes, at least until Apple breaks things on us again.

**Comment by akhansen** [comment]
@mrvacbob, @nieder, that's probably due to what @PovlAbrahamsen noted above, because when the test fails it falls back to 10.14 instead of 10.14.5.

File: perlmod/Fink/Bootstrap.pm (hunk 1 above)
Comment (nieder) [reply perlmod/Fink/Bootstrap.pm]: fixed in 648179b

**Comment by nieder** [comment]
648179b should take care of unreleased versions accidentally falling back to Dist: 10.14 and not 10.14.5.

However, I think line [260 in Engine.pm](https://github.com/fink/fink/pull/190/files#diff-1e8680eec65e822fc7191589c3b20c85R260) needs to be tweaked because I think I hardcoded it to only work as intended when 10.14.5 (distribution field) == 10.14.5 (os version). And I don't know enough perl to grok the versions correctly as version numbers and not as strings or decimals.

**Comment by akhansen** [comment]
version->parse() is what is used to convert the strings/decimals into version strings.

**Comment by nieder** [comment]
That conditional is driving me nuts.
Don't do an upgrade if 
1) $osversion == $distribution (<=10.14 and presumably 10.15 coming up if Apple returns to form), or
2) if $distribution == 10.14.5, then check if $osversion_long == $distribution.

(2) is meant to catch macOS 10.14.[67...], where $distribution = 10.14.5, but $osversion_long != 10.14.5. And the leading if for (2) is because we normally don't want to compare against $osversion_long, except in the 10.14.5 case.

Does this make sense?

**Comment by nieder** [comment]
To update, this worked for me as a both a fresh install (./bootstrap) on 10.14.5 and just tested an upgrade from 0.44.1 (./inject) on 10.14.5.

Not tested on 10.14.6 but I think the pseudocode in m [previous comment](https://github.com/fink/fink/pull/190#issuecomment-499025909) should work if someone could help with the actual perl code.

**Comment by hanbyul1** [comment]
I am trying to install fig, but got stuck with the following message. It was a fresh install of fink, which went well with the shared solutions. The selfupdate was went well too, but got stuck at the last final step. Thanks for any comments!

***************$ /sw/bin/fink install xfig
Information about 10628 packages read in 1 seconds.
Can't resolve dependency "texi2html" for package "autoconf2.6-2.69-6" (no matching packages/versions found)
Exiting with failure.

**Comment by akhansen** [comment]
Haven

**Comment by akhansen** [comment]
I looked at this a bit, but I haven't gotten to the point of actual code yet--I've been doing paid job stuff over the weekend.

File: perlmod/Fink/Bootstrap.pm (hunk 1 above)
Comment (huddini79) [reply perlmod/Fink/Bootstrap.pm]: I am a n00b, so what do I have to do to install fink on fresh mac with 10.14.5 .... cheers

**Comment by nieder** [comment]
To install this work in progress Fink on a fresh Mac with 10.14.5

* Download https://github.com/fink/fink/archive/nieder-5.18.4-bootstrap.zip
* expand zip file, and from inside directory, run `./bootstrap`.

**Comment by akhansen** [comment]
Where are we at on this?

**Comment by nieder** [comment]
I've tested fresh installs on 10.14.5 and .6beta and updates from a previous 10.14.4 setup. Other people on the mailing lists have also reported success.

**Comment by sclint13** [comment]
Hey, I have been trying to download Fink to my Mac on 10.14.5 using the zip from nieder, but I keep getting an error with my dpkg files. I have attached the latest log from running bootstrap. I also tried `fink selfupdate` but it fails saying 

> Scanning package description files..........
> Information about 10640 packages read in 16 seconds.
> The following 3 packages will be installed or updated:
>  dpkg dpkg-base-files tar
> Setting runtime build-lock...
> dpkg-deb -b /sw/src/fink.build/root-fink-buildlock-dpkg-1.10.21-1246 /sw/src/fink.build
> Can't exec "dpkg-deb": No such file or directory
> ### execution of dpkg-deb failed, exit code -1
> Failed: can't create package fink-buildlock-dpkg-1.10.21-1246

It says this whether I use git or rsync. I've tried searching the Mailing List but I thought I may have better luck on here in case it is something to do with nieder's compilation?
[Bootstrap.txt](https://github.com/fink/fink/files/3323265/Bootstrap.txt)

**Comment by PovlAbrahamsen** [comment]
@sclint13, the issue isn't with this patch, but with a separate copy of libz, which you appear to have installed in /usr/local/lib. This seems to be throwing off the linker, preventing it from linking to the system-installed libz in /usr/lib.

> gcc -L/sw/lib -framework CoreFoundation  -o dpkg-deb  build.o  extract.o  info.o  main.o ../lib/libdpkg.a ../optlib/libopt.a -lz -lintl 
> ld: warning: ignoring file /usr/local/lib/libz.a, file was built for archive which is not the architecture being linked (x86_64): /usr/local/lib/libz.a

**Comment by nieder** [comment]
Sorry, I pushed a couple more changes. I think I'm done doing that.

**Comment by jaokiro** [comment]
I am a beginner to coding and was trying to install fink on a Mac OS 10.14.5 where I stumbled on the Perl error. I therefore used https://github.com/fink/fink/archive/nieder-5.18.4-bootstrap.zip and the installation was complete. However, when trying to launch fink, I get this error:

_You should now have a working Fink installation in '/sw'. You still need package descriptions if you want
to compile packages yourself. You can get them by running either of the commands: 'fink selfupdate-rsync',
to update via rsync (generally preferred); or 'fink selfupdate-svn', to update via svn (more likely to work
through a firewall). You should also run 'sudo xcodebuild -license' and accept the Xcode license, since
some packages will require this.

Run '. /sw/bin/init.sh' to set up this terminal session environment to use Fink. To make the software
installed by Fink available in all of your future terminal shells, add '. /sw/bin/init.sh' to the init
script '.profile' or '.bash_profile' in your home directory. The program /sw/bin/pathsetup.sh can help with
this. Enjoy.

Jareds-MacBook-Pro:fink-nieder-5.18.4-bootstrap jaredokiro$  /sw/bin/init.sh
-bash: /sw/bin/init.sh: Permission denied
Jareds-MacBook-Pro:fink-nieder-5.18.4-bootstrap jaredokiro$ cd
Jareds-MacBook-Pro:~ jaredokiro$  /sw/bin/init.sh
-bash: /sw/bin/init.sh: Permission denied
Jareds-MacBook-Pro:~ jaredokiro$ 

Kindly assist. Thanks._

**Comment by PovlAbrahamsen** [comment]
> Run '. /sw/bin/init.sh' to set up this terminal session environment to use Fink. To make the software
> installed by Fink available in all of your future terminal shells, add '. /sw/bin/init.sh' to the init
> script '.profile' or '.bash_profile' in your home directory. The program /sw/bin/pathsetup.sh can help with
> this. Enjoy.
> 
> Jareds-MacBook-Pro:fink-nieder-5.18.4-bootstrap jaredokiro$ /sw/bin/init.sh
> -bash: /sw/bin/init.sh: Permission denied
> Jareds-MacBook-Pro:fink-nieder-5.18.4-bootstrap jaredokiro$ cd
> Jareds-MacBook-Pro:~ jaredokiro$ /sw/bin/init.sh
> -bash: /sw/bin/init.sh: Permission denied

You need to run ". /sw/bin/init.sh" - note the period and space before "/sw/bin/init.sh".
Or, alternatively, run /sw/bin/pathsetup.sh (without a period and space) to set up your computer to use Fink.

**Comment by jaokiro** [comment]
Thanks. It works now:)

**Comment by peytondh** [comment]
Not all seems perfect, as shown by trying to install package:
"molmol              2k.2.0-1056     Molecular graphics display program"

So..., as pasted from the terminal under tcsh:
% fink install molmol
Information about 10644 packages read in 1 seconds.
Can't resolve dependency "texi2html" for package "autoconf2.6-2.69-6" (no matching
packages/versions found)
Exiting with failure.

**Comment by akhansen** [comment]
@peytondh that's an independent issue.

**Comment by peytondh** [comment]
akhansen:
Sure, but it seems that there are multiple such definition files missing after a self-update. So a lot of installs can't be done. These were not missing in earlier versions.

**Comment by akhansen** [comment]
My point is that those have nothing to do with the fink code itself.  Those are covered in the Fink distribution repository (fink/fink-distributions).  Basically, nothing that goes on here in this repository will address that.

**Comment by peytondh** [comment]
Thanks, and forgive my ignorance about devision of responsibilities. I just wanted to let people know about the issue, so that the right people could address it.

**Comment by nieder** [comment]
@peytondh Package changes for 10.14.5 are tracked in fink/fink-distributions#423

**Comment by peytondh** [comment]
Confirmed. Works.

Thanks to all!

**Pull request merged**